### ISSUE TITLE
Workshop on multiprocessing

### ISSUE THREAD

**Issue by Ludee** [OpenEnergyPlatform/open-MaStR#79]
Hi @Bachibouzouk 
I struggle with the code for the multiprocessing for the download.
More download functions need a speed up with stable retry options.
I would like to develop a template I can use in this repo.
A refresh on the test would be nice as well.

- [tests/test_parallel_download.py](https://github.com/OpenEnergyPlatform/open-MaStR/blob/master/tests/test_parallel_download.py)
- [download_parallel_power_unit()](https://github.com/OpenEnergyPlatform/open-MaStR/blob/master/soap_api/mastr_power_unit_download.py#L144-L323)
- [download_parallel_unit_solar()](https://github.com/OpenEnergyPlatform/open-MaStR/blob/master/soap_api/mastr_solar_download.py#L135-L187)
- [download_parallel_unit_solar_eeg()](https://github.com/OpenEnergyPlatform/open-MaStR/blob/master/soap_api/mastr_solar_download.py#L353-L404)

**Comment by Bachibouzouk** [comment]
```
import multiprocessing as mp


def f(x):
    print("recieving", x)
    return x * x


mylist = []


def append_result(res):
    print(res)
    mylist.append(res)


if __name__ == '__main__':

    pool = mp.get_context("spawn").Pool(processes=3)

    sublist = [1, 2, 3]

    r = pool.map_async(f, sublist, callback=append_result)

    pool.close()
    pool.join()
    pool.terminate()

    r.wait()

    print(mylist)

```

The idea is to split the call for the download (`c = client_bind.GetEinheitSolar`, `f` in my example) and the process of the downloaded data (which can be defined in the callback, `append_result` in my example)

**Comment by jadeaffenjaeger** [comment]
https://github.com/OpenEnergyPlatform/open-MaStR/pull/99 moves all parallelization to the single function [parallel()](https://github.com/jadeaffenjaeger/open-MaStR/blob/a674aefbaf94b403ec712abd3368257cd161dfdb/soap_api/parallel.py#L33-L50). This function takes a list of candidate identifiers and a function that will be mapped to them in parallel as an input. See [download_parallel_unit_solar(...)](https://github.com/jadeaffenjaeger/open-MaStR/blob/a674aefbaf94b403ec712abd3368257cd161dfdb/soap_api/mastr_solar_download.py#L132-L133) for an example of what this looks like.

This keeps the parallelization in a single place and should make improvements/maintenance easier.
The function [imap_unordered](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.pool.Pool.imap_unordered) is used here because different from `map`/`map_async`, it allows to iterate over results that were produced already, while `map` will block until *all* threads have finished *all* of their work. Through this, the parent thread can consume already downloaded units and write them to the corresponding CSV file. This also keeps the file access in a single thread and avoids problems with concurrent file accesses from multiple threads. See https://github.com/OpenEnergyPlatform/open-MaStR/issues/95#issuecomment-685435723 for details.

The division of the candidate lists into sublists before passing it on to the threads is also removed. Unfortunately, I don't know what the rationale for picking this approach might have been. I would assume that the same effect can be achieved if necessary by varying the `chunk_size` parameter in the `multiprocessing` API!? Parallel download seems to work without problems with the default value as far as I can tell this far, though.

**Comment by gplssm** [comment]
Closed by #99
